python-frontend/src/test/java/org/sonar/python/semantic/v2/converter/PythonTypeToDescriptorConverterTest.java

```diff
@@ -0,0 +1,226 @@
+    assertThat(ambiguousDescriptor.fullyQualifiedName()).isEqualTo("foo.myUnionType");
+    assertThat(ambiguousDescriptor.kind()).isEqualTo(Descriptor.Kind.AMBIGUOUS);
+    // TODO SONARPY-2225 the two class types in the union are rigorously the same but the converter creates an ambigouous symbol
+    assertThat(ambiguousDescriptor.alternatives()).hasSize(2);
+    assertThat(ambiguousDescriptor.alternatives()).extracting(Descriptor::name).containsExactlyInAnyOrder("myUnionType", "myUnionType");
+    assertThat(ambiguousDescriptor.alternatives()).extracting(Object::getClass).allMatch(c -> c == ClassDescriptor.class);
+  }
+
+  @Test
+  void testConvertManyTypes() {
+    ClassType classType = new ClassType("classType", Set.of(new Member("aMember", intTypeWrapper.type())), List.of(), List.of(floatTypeWrapper), List.of(intTypeWrapper.type()), true, location);
+    FunctionType functionType = new FunctionType("functionType", List.of(new ModuleType("bar")), List.of(), floatTypeWrapper, TypeOrigin.LOCAL, true, false, true, false, null, location);
+    Descriptor descriptor = converter.convert("foo", new SymbolV2("myUnionType"), Set.of(functionType, classType));
+
+    assertThat(descriptor).isInstanceOf(AmbiguousDescriptor.class);
+    AmbiguousDescriptor ambiguousDescriptor = (AmbiguousDescriptor) descriptor;
+    assertThat(ambiguousDescriptor.name()).isEqualTo("myUnionType");
+    assertThat(ambiguousDescriptor.fullyQualifiedName()).isEqualTo("foo.myUnionType");
+    assertThat(ambiguousDescriptor.kind()).isEqualTo(Descriptor.Kind.AMBIGUOUS);
+    assertThat(ambiguousDescriptor.alternatives()).hasSize(2);
+    assertThat(ambiguousDescriptor.alternatives()).extracting(Descriptor::kind).containsExactlyInAnyOrder(Descriptor.Kind.CLASS, Descriptor.Kind.FUNCTION);
+  }
+
+  @Test
+  void testConvertManyTypesWithUnionType() {
+    ClassType classType = new ClassType("classType", Set.of(new Member("aMember", intTypeWrapper.type())), List.of(), List.of(floatTypeWrapper), List.of(intTypeWrapper.type()), true, location);
+    ClassType anotherClassType = new ClassType("classType", Set.of(new Member("aMember", intTypeWrapper.type())), List.of(), List.of(floatTypeWrapper), List.of(intTypeWrapper.type()), true, location);
+
+    UnionType unionType = new UnionType(Set.of(classType, anotherClassType));
+    Descriptor descriptor = converter.convert("foo", new SymbolV2("myUnionType"), Set.of(unionType, classType));
+
+    assertThat(descriptor).isInstanceOf(AmbiguousDescriptor.class);
+    AmbiguousDescriptor ambiguousDescriptor = (AmbiguousDescriptor) descriptor;
+    assertThat(ambiguousDescriptor.name()).isEqualTo("myUnionType");
+    assertThat(ambiguousDescriptor.fullyQualifiedName()).isEqualTo("foo.myUnionType");
+    assertThat(ambiguousDescriptor.kind()).isEqualTo(Descriptor.Kind.AMBIGUOUS);
+    // TODO SONARPY-2225 the two class types in the union are rigorously the same but the converter creates an ambigouous descriptor
```

Comment:
I believe it is to be expected since we don't implement `equals` for `PythonType` and expect the unicity of PythonType objects for each available type.

I guess that means that we'd have an ambiguous descriptor in the case where there are 2 definitions that are strictly identical:
```
if x:
  class A: ...
else:
  class A: ...
```

However, I think it's okay to accept this behavior for now. If this prevents proper resolution in some cases, we can always implement something that would merge types that are essentially identical.